types/evidence_test.go

```diff
@@ -460,3 +459,107 @@ func makeHeaderRandom() *Header {
 		ProposerAddress:    crypto.CRandBytes(crypto.AddressSize),
 	}
 }
+
+func TestEvidenceProto(t *testing.T) {
+	// -------- Votes --------
+	val := NewMockPV()
+	blockID := makeBlockID(tmhash.Sum([]byte("blockhash")), math.MaxInt64, tmhash.Sum([]byte("partshash")))
+	blockID2 := makeBlockID(tmhash.Sum([]byte("blockhash2")), math.MaxInt64, tmhash.Sum([]byte("partshash")))
+	const chainID = "mychain"
+	v := makeVote(t, val, chainID, math.MaxInt64, math.MaxInt64, 1, 0x01, blockID)
+	v2 := makeVote(t, val, chainID, math.MaxInt64, math.MaxInt64, 2, 0x01, blockID2)
+
+	// -------- SignedHeaders --------
+	const height int64 = 37
+
+	var (
+		header1 = makeHeaderRandom()
+		header2 = makeHeaderRandom()
+	)
+
+	header1.Height = height
+	header1.LastBlockID = blockID
+	header1.ChainID = chainID
+
+	header2.Height = height
+	header2.LastBlockID = blockID
+	header2.ChainID = chainID
+
+	voteSet1, valSet, vals := randVoteSet(height, 1, PrecommitType, 10, 1)
+	voteSet2 := NewVoteSet(chainID, height, 1, PrecommitType, valSet)
+
+	commit1, err := MakeCommit(BlockID{
+		Hash: header1.Hash(),
+		PartsHeader: PartSetHeader{
+			Total: 100,
+			Hash:  crypto.CRandBytes(tmhash.Size),
+		},
+	}, height, 1, voteSet1, vals, time.Now())
+	require.NoError(t, err)
+	commit2, err := MakeCommit(BlockID{
+		Hash: header2.Hash(),
+		PartsHeader: PartSetHeader{
+			Total: 100,
+			Hash:  crypto.CRandBytes(tmhash.Size),
+		},
+	}, height, 1, voteSet2, vals, time.Now())
+	require.NoError(t, err)
+
+	h1 := &SignedHeader{
+		Header: header1,
+		Commit: commit1,
+	}
+	h2 := &SignedHeader{
+		Header: header2,
+		Commit: commit2,
+	}
+
+	tests := []struct {
+		testName string
+		evidence Evidence
+		wantErr  bool
+		wantErr2 bool
+	}{
+		{"&DuplicateVoteEvidence empty fail", &DuplicateVoteEvidence{}, true, true},
+		{"&DuplicateVoteEvidence nil voteB", &DuplicateVoteEvidence{VoteA: v, VoteB: nil}, true, true},
+		{"&DuplicateVoteEvidence nil voteA", &DuplicateVoteEvidence{VoteA: nil, VoteB: v}, true, true},
+		{"&DuplicateVoteEvidence success", &DuplicateVoteEvidence{VoteA: v2, VoteB: v}, false, false},
+		{"&ConflictingHeadersEvidence empty fail", &ConflictingHeadersEvidence{}, true, true},
+		{"&ConflictingHeadersEvidence nil H2", &ConflictingHeadersEvidence{H1: h1, H2: nil}, true, true},
+		{"&ConflictingHeadersEvidence nil H1", &ConflictingHeadersEvidence{H1: nil, H2: h2}, true, true},
+		// {"&ConflictingHeadersEvidence success", &ConflictingHeadersEvidence{H1: h1, H2: h2}, false, false},
```

Comment:
Everytime evidence goes to proto it will come back a concrete type instead of a pointer. Is this okay? I dont see a issue with this but I dont understand the evidence part that well.

cc @melekes @cmwaters

Reply:
concrete types are fine